src/shared/utilities/messages.ts

```diff
@@ -182,11 +182,134 @@ export async function showMessageWithCancel(
     message: string,
     timeout: Timeout,
     window: Window = globals.window
-): Promise<vscode.Progress<{ message?: string; increment?: number }>> {
+): Promise<vscode.Progress<ProgressEntry>> {
     const progressOptions = { location: vscode.ProgressLocation.Notification, title: message, cancellable: true }
     return showProgressWithTimeout(progressOptions, timeout, window)
 }
 
+/**
+ * A timeout that is intended to be used by
+ * {@link CancellableMessageManager} so that the manager
+ * can perform cleanup when the timeout is stopped.
+ */
+class CancellableMessageTimeout extends Timeout {
+    constructor(timeoutLength: number, readonly msgId: string) {
+        super(timeoutLength)
+    }
+
+    public override cancel(): void {
+        super.cancel()
+        CancellableMessageManager.cancelMessage(this.msgId)
+    }
+
+    public override dispose(): void {
+        super.dispose()
+        CancellableMessageManager.disposeMessage(this.msgId)
+    }
+}
+
+/**
+ * Provides support to create cancellable vscode message windows.
+ *
+ * To get started, you create a new message with a unique ID.
+ * Following that, you can easily update the message or cancel it.
+ *
+ * ---
+ *
+ * This class solves an issue of needing to manage duplicated
+ * messages, the interface simplifies the process.
+ */
+export class CancellableMessageManager {
```

Comment:
Can avoid "premature explanation" here.

```suggestion
export class MessageManager {
```